### DISCUSSION TITLE
What is the history behind the name Bind & Tap?

### DISCUSSION THREAD

**Discussion by xavierjohn** [vkhorikov/CSharpFunctionalExtensions#486] [Q&A]
What is the history behind the names Bind & Tap?
The names do not express intent.

**Comment by xavierjohn** [comment]
If there is no specific reason for Bind and Tap, I think there is no need of Tap, Bind can deal with both scenarios.
Bind can be renamed to IfOk and we will get something like this.

```csharp
 var result = await GetCustomerByIdAsync(id)
       .ToResultAsync(Error.NotFound("Customer with such Id is not found: " + id))
       .EnsureAsync(customer => customer.CanBePromoted, Error.Validation("Need to ask manager"))
       .IfErrorAsync(error => Log(error))
       .IfErrorAsync(() => AskManagerAsync(id))
       .IfOkAsync(customer => Log("Manager approved promotion"))
       .IfOkAsync(customer => customer.PromoteAsync())
       .IfOkAsync(customer => EmailGateway.SendPromotionNotificationAsync(customer.Email))
       .UnwrapAsync(ok => "Okay", error => error.Message);
```

Opinions?

**Comment by vkhorikov** [comment]
The names are from F# (and other FP languages) and TypeScript respectively. This library tries to avoid introducing its own terminology unless absolutely necessary. But I agree, `IfSuccess` / `IfError` do look more readable.

**Comment by xavierjohn** [comment]
My friends who are familiar with F# have not heard of Tap. 
What is the name of these libraries?

I found one article but it uses, map, then and join
https://itnext.io/railroad-programming-in-typescript-21d69f486f6e

**Comment by xavierjohn** [comment]
Are the names based on https://fsharpforfunandprofit.com/rop/  ?

  Using a list of custom error types on both the left and right sides of Either (rather than, say, Either String a).
    “bind” (>>=) for integrating monadic functions into the pipeline.
    Kleisli composition (>=>) for composing monadic functions.
    “map” (fmap) for integrating non-monadic functions into the pipeline.
    “tee” for integrating unit functions into the pipeline (because F# doesn’t use the IO monad).
    Mapping from exceptions to error cases.
    &&& for combining monadic functions in parallel (e.g. for validation).
    The benefits of custom error types for domain driven design.

**Comment by vkhorikov** [comment]
`Tap` is from TypeScript, yes.

> Are the names based on https://fsharpforfunandprofit.com/rop/ ?

Correct.

**Comment by geniuskld** [comment]
Я заметил есть tap, check & bind. Где-то можно почитать о специфике и различиях?
